### PR TITLE
Refactor&Feature: Rebuild rapidjson module.

### LINKED ISSUE #3539: Feature: Json formatted output is required.
### Background

At present, there are many other software or applications that implement interfaces with ABACUS for some post-processing, such as dpgen/ASE/atom-kit/abacustest. The main parameters of ABACUS or the output of some key values are mostly saved in the OUT.XXX/running_YYY.log file. 
The extraction of information from this text file is inefficient and unstable, and often faces the problem that the interface cannot be used due to the modification of the output format, so we need a formatted data storage file.

### Describe the solution you'd like

Goal: To facilitate post-processing software to read the results. (such as dpgen/abacustest/phonopy, etc.) Heavy content is not added (such as track coefficient and other special format files)
- Hierarchy: The hierarchy is minimized to facilitate subsequent values.
- Output frequency: Maintaining high output frequency may affect computing performance.
- Output accuracy: direct output without reducing accuracy (need to query rapidjson for output Settings)
- Add an installation library rapidjson to increase the user threshold: add compilation options, turn off the output by default, and open compilation options to install if the user needs this function.
- Use: Add documentation that explains to the user how to use python to call this file and how key information should be obtained.
Output path: INPUT peer path

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

_No response_

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### PR DESCRIPTION
**Pull request by grysgreat**
Construct json output with rapidjson in a more concise way

### Linked Issue
Fix #3539

### What's changed?
- codes connected to json output

### CODE REVIEW

File: source/module_base/parallel_global.cpp [hunk 1]
```diff
@@ -26,7 +26,8 @@ MPI_Comm STO_WORLD;
 MPI_Comm PARAPW_WORLD; // qianrui add it for sto-dft 2021-4-14
 MPI_Comm GRID_WORLD; // mohan add 2012-01-13z
 MPI_Comm DIAG_WORLD; // mohan add 2012-01-13
-
+int mpi_number=0;
+int omp_number=0;
```
Comment (pxlxingliang) [inline source/module_base/parallel_global.cpp:32]: Should assign mpi_number and omp_number as the real value

File: source/module_io/json_output/test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+remove_definitions(-D__DEEPKS)
+remove_definitions(-D__CUDA)
+remove_definitions(-D__ROCM)
+remove_definitions(-D__EXX)
+remove_definitions(-DUSE_PAW)
+
+AddTest(
+  TARGET io_json_output_json
+  SOURCES ../json.cpp
```
Comment (pxlxingliang) [inline source/module_io/json_output/test/CMakeLists.txt]: should be ../abacusjson.cpp?

**Comment by caic99** [comment]
@grysgreat Please consider making the option default to on with docs added. Otherwise, please enable this feature only for necessary tests.

File: source/module_io/json_output/test/para_json_test.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    Json::AbacusJson::doc.Parse("{}");
+    Json::json_output();
+
+    std::string filename = "abacus.json";
+    std::ifstream file(filename);
+    ASSERT_TRUE(file.is_open());
+
+    std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+
+    ASSERT_NE(content.find("\"version\": \"v3.5.1\","), std::string::npos);
```
Comment (pxlxingliang) [inline source/module_io/json_output/test/para_json_test.cpp]: Do not check the value of version, which is different in different version.

File: source/module_io/json_output/test/para_json_test.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    ASSERT_NE(content.find("\"device\": \"cpu\","), std::string::npos);
+    ASSERT_NE(content.find("\"omp_num\": 0,"), std::string::npos);
```
Comment (pxlxingliang) [inline source/module_io/json_output/test/para_json_test.cpp:98]: The value of omp_num should not be 0

File: source/module_io/json_output/test/para_json_test.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    ASSERT_NE(content.find("\"key1\": \"value1\","), std::string::npos);
+    ASSERT_NE(content.find("\"key2\": {"), std::string::npos);
+    ASSERT_NE(content.find("\"key3\": 1"), std::string::npos);
+    ASSERT_NE(content.find("\"key4\": 0.1"), std::string::npos);
+    ASSERT_NE(content.find("\"key5\": true"), std::string::npos);
+
+    file.close();
+}
+
+TEST(AbacusJsonTest, GeneralInfo)
```
Comment (pxlxingliang) [inline source/module_io/json_output/test/para_json_test.cpp:74]: This test should check all general_info parameters

**Review by pxlxingliang — APPROVED** [review]

**Pull request merged**